WordPress/gutenberg
# Nav block: enable creation from existing WP Menus

<img width="647" alt="Screenshot 2020-05-29 at 15 40 38" src="https://user-images.githubusercontent.com/444434/83272216-cdff2800-a1c2-11ea-9cdc-d2bee79ccbf6.png">

This PR allow the creation of Nav Block (items) from existing WP Menus. ~Requires[ the experimental Menu REST API Plugin](https://github.com/WP-API/menus-endpoints). The intent is that this endpoint gets merged into WordPress Core thereby enabling this functionality on the Block to become a reality.~ (**update**: the endpoints are now merged into Gutenberg Core).

This is visually based on [this mockup from @shaunandrews](https://github.com/WordPress/gutenberg/issues/18575#issuecomment-579896803).

**Note:** the creation of the Nav Block from the Menu is a one time action. It does not keep the Nav Block items in sync with the Menu.

~Depends on #20292~
Closes https://github.com/WordPress/gutenberg/issues/18575

## Todo

The following items are outstanding on this PR:

- [x] Update e2e tests to match new design.
- [x] Fix or workaround Axe accessibility tests errors in component dependencies.
- [x] Fix or work around `React state update on an unmounted component` relating to `ToolbarContainer` component (again not modified in this PR so unrelated???).
- [ ] Fix CSS logic to ensure the dividing line is always below the last "Menu" item. We cannot rely on there always being a "Create from pages" option. See https://d.pr/i/qfvMbT.


## Testing Instructions

If you notice any "lag" on inserting/selecting the Block you are probably experiencing https://github.com/WordPress/gutenberg/issues/22823.

Does this match [the approved design](https://github.com/WordPress/gutenberg/pull/18869#issuecomment-622396161)?

### Pre-test setup

* Remove all Menus (Appearance -> Menus) from your WP site.
* Switch to this branch. Rebuild. Any issues with Prettier please [see here](https://wordpress.slack.com/archives/C02QB2JS7/p1584643823052800)).

### Test Flow

* Create a New Post
* Add Nav Block
* You should **_not_** see any Menu options in the placeholder dropdown as there are no menus.
* Goto Apperance -> Menus and create a new Menu but **do not add any items to it**.
* Create a New Post
* Add Nav Block
* You _should_ see the Menu you created within the dropdown. Select the Menu from the dropdown and click the "Create" button.
* You should see an empty Nav Block created with no Nav Item Blocks - this reflects the empty state of your WP Menu.
* Goto Appearance -> Menus and add some items to your Menu (be sure to hit "Save" to save your Menu else it will still have 0 items!).
* Create a New Post
* Add Nav Block
* You _should_ still see the Menu you created within the dropdown. Select the Menu from the dropdown and click the "Create" button.
* You should see the Nav Block created with all of the items that exist in your WP Menu.

* Test that you can still create from **Pages** and create an **empty** Nav block.
* Also try the tests above on a limited network connection (simulate using Devtools "Network" tab).

## How has this been tested?


## Screenshots 

![Screen Capture on 2020-05-29 at 15-40-12](https://user-images.githubusercontent.com/444434/83272288-deaf9e00-a1c2-11ea-8561-723adf26649b.gif)


## Types of changes
New feature (non-breaking change which adds functionality) 

## Checklist:
